File: backend/app/api/v1/endpoints/statarb.py
"""
Statistical Arbitrage API endpoints.
"""
from typing import List
from fastapi import APIRouter, Depends, BackgroundTasks, HTTPException, Query
from sqlmodel import Session, select
from app.core.db import get_session, engine
from app.models.statarb import CointegratedPair
from app.services.cointegration import CointegrationService
from app.schemas.statarb import AnalysisRequest, AnalysisResponse, SpreadPoint

router = APIRouter(prefix="/statarb", tags=["StatArb"])

def run_analysis_task(universe: List[str]):
    """Background task to run pairs analysis."""
    with Session(engine) as session:
        service = CointegrationService()
        service.find_pairs(universe, session)

@router.post("/analyze", response_model=AnalysisResponse)
async def analyze_pairs(
    request: AnalysisRequest, 
    background_tasks: BackgroundTasks
):
    """
    Start a cointegration analysis on the provided universe.
    Runs in the background.
    """
    background_tasks.add_task(run_analysis_task, request.universe)
    return {"message": f"Analysis started for {len(request.universe)} tickers", "status": "processing"}

@router.get("/pairs", response_model=List[CointegratedPair])
async def get_pairs(session: Session = Depends(get_session)):
    """Get list of active cointegrated pairs."""
    return session.exec(select(CointegratedPair).where(CointegratedPair.is_active == True)).all()

@router.get("/spread", response_model=List[SpreadPoint])
async def get_spread(
    ticker1: str = Query(...),
    ticker2: str = Query(...),
    session: Session = Depends(get_session)
):
    """Get historical Z-Score spread for visualization."""
    service = CointegrationService()
    data = service.get_spread_series(ticker1, ticker2, session)
    if not data:
        raise HTTPException(status_code=404, detail="Insufficient data for pair")
    return data

File: frontend/src/services/statarb.ts
/**
 * StatArb API service.
 *
 * Client for accessing pairs trading analysis and results.
 */

import { api } from "@/lib/api";

// ==============================================================================
// Types
// ==============================================================================

export interface CointegratedPair {
    id: number;
    ticker_1: string;
    ticker_2: string;
    p_value: number;
    hedge_ratio: number;
    half_life: number;
    last_z_score: number;
    is_active: boolean;
}

export interface AnalysisRequest {
    universe: string[];
}

export interface AnalysisResponse {
    message: string;
    status: string;
}

export interface SpreadPoint {
    date: string;
    z_score: number;
}

// ==============================================================================
// API Functions
// ==============================================================================

/**
 * Trigger pairs analysis.
 */
export async function analyzePairs(universe: string[]): Promise<AnalysisResponse> {
    const response = await api.post<AnalysisResponse>("/statarb/analyze", { universe });
    return response.data;
}

/**
 * Get active cointegrated pairs.
 */
export async function getPairs(): Promise<CointegratedPair[]> {
    const response = await api.get<CointegratedPair[]>("/statarb/pairs");
    return response.data;
}

/**
 * Get z-score spread history for visualization.
 */
export async function getSpread(ticker1: string, ticker2: string): Promise<SpreadPoint[]> {
    const response = await api.get<SpreadPoint[]>("/statarb/spread", {
        params: { ticker1, ticker2 }
    });
    return response.data;
}
